pkg/products/threescale/reconciler.go

```diff
@@ -342,6 +383,260 @@ func (r *Reconciler) backupSystemSecrets(ctx context.Context, serverClient k8scl
 	return integreatlyv1alpha1.PhaseCompleted, nil
 }
 
+func getDeploymentConfig(ctx context.Context, client k8sclient.Client, dcName string) (*appsv1.DeploymentConfig, integreatlyv1alpha1.StatusPhase, error) {
+	apiCastDeploymentConfig := &appsv1.DeploymentConfig{}
+
+	err := client.Get(ctx, k8sTypes.NamespacedName{Name: dcName, Namespace: "redhat-rhmi-3scale"}, apiCastDeploymentConfig)
+
+	if err != nil {
+		if k8serr.IsNotFound(err) {
+			return nil, integreatlyv1alpha1.PhaseAwaitingComponents, nil
+		}
+		return nil, integreatlyv1alpha1.PhaseFailed, fmt.Errorf("error getting the deployment config %v", err)
+	}
+	return apiCastDeploymentConfig, integreatlyv1alpha1.PhaseInProgress, nil
+}
+
+func getService(ctx context.Context, client k8sclient.Client, dcName string) (*corev1.Service, integreatlyv1alpha1.StatusPhase, error) {
+	service := &corev1.Service{}
+
+	err := client.Get(ctx, k8sTypes.NamespacedName{Name: dcName, Namespace: "redhat-rhmi-3scale"}, service)
+
+	if err != nil {
+		if k8serr.IsNotFound(err) {
+			return nil, integreatlyv1alpha1.PhaseAwaitingComponents, nil
+		}
+		return nil, integreatlyv1alpha1.PhaseFailed, fmt.Errorf("error getting the deployment config %v", err)
+	}
+	return service, integreatlyv1alpha1.PhaseInProgress, nil
+
+}
+
+func (r *Reconciler) addEnvoyConfig(ctx context.Context, client k8sclient.Client, deploymentConfig *appsv1.DeploymentConfig, dcName string, service *corev1.Service, namespace string) error {
+
+	secretName := fmt.Sprintf("%s-certificate", dcName)
+	_, err := controllerutil.CreateOrUpdate(ctx, client, service, func() error {
+
+		if service.Annotations == nil {
+			service.Annotations = map[string]string{}
+		}
+		// add annotation in order to trigger openshift to create a required tls secret
+		service.Annotations["service.beta.openshift.io/serving-cert-secret-name"] = secretName
+
+		// update the gateway port to point to 8443 so that the request is via the envoy sidecar container
+
+		ports := service.Spec.Ports
+		for i, port := range ports {
+			if port.Name == "gateway" {
+				logrus.Info("updating the gateway port")
+				service.Spec.Ports[i].Port = 8443
+				service.Spec.Ports[i].TargetPort = intstr.FromInt(8443)
+				break
+			}
+		}
+		return nil
+	})
```

Comment:
```suggestion
if service.Annotations == nil {
	service.Annotations = map[string]string{}
}
// add annotation in order to trigger openshift to create a required tls secret
service.Annotations["service.beta.openshift.io/serving-cert-secret-name"] = secretName

// update the gateway port to point to 8443 so that the request is via the envoy sidecar container

ports := service.Spec.Ports
for i, port := range ports {
	if port.Name == "gateway" {
		logrus.Info("updating the gateway port")
		service.Spec.Ports[i].Port = 8443
		service.Spec.Ports[i].TargetPort = intstr.FromInt(8443)
		break
	}
}

if err := client.Update(ctx, service); err != nil {
    return fmt.Errorf("failed to update service: %v", err)
}
```

There's no need to use the `CreateOrUpdate` function as this code won't be reached if the service doesn't exist (it will have returned `AwaitingComponents`). Same thing can be changed for the DeploymentConfig